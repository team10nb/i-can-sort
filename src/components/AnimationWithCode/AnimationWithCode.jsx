/*
    Author: Ruizi Han, Yijie Lu

    Framework of animation and pseudocode, support for 2 loop examples and Terminating Bubble Sort in correctness-tutorial-termination.
*/

import "@fontsource/roboto";
import * as React from "react";
import { motion } from "framer-motion";
import { useState, useEffect } from "react";
import AnimationSlider from "../AnimationSlider/AnimationSlider";
import AnimationControl from "../AnimationControl/AnimationControl";
import ArrowDropUpIcon from '@material-ui/icons/ArrowDropUp';
import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';


// a framer motion transition attributes
const spring = {
    type: "spring", // a framer motion type that simulates spring
    damping: 15, //Strength of opposing force. If set to 0, spring will oscillate indefinitely
    stiffness: 150, //Stiffness of the spring. Higher values will create more sudden movement. Set to 100 by default.
    mass: 0.1, // Mass of the moving object. Higher values will result in more lethargic movement
};

export default function AnimationWithCode(props) {
    const { trace, description, width, blockNums, useStyles, Code, title } = props;

    // The bars displayed to visulise the numbers
    const [bars, setBars] = useState(trace[0]);
    // The speed of playing the animation
    const [playSpeed, setPlaySpeed] = useState(1);
    // The current step among traces
    const [currentStep, setCurrentStep] = useState(0);
    // It is used to clean timeouts to pause
    const [timeOutIds, setTimeOutIds] = useState([]);
    // The state of the animation, whether is playing
    const [isPlaying, setIsPlaying] = useState(false);
    // It is used to control the speed menu
    const [anchorEl, setAnchorEl] = useState(null);
    // The state of play button and next step button, whether is disabled
    const [playDisabled, setPlayDisabled] = useState(false);
    // the State of previous step button
    const [backwardDisabled, setBackwardDisabled] = useState(true);

    const classes = useStyles();

    // Update buttons' disable property when steps are changed
    useEffect(() => {
        currentStep === 0
            ? setBackwardDisabled(true)
            : setBackwardDisabled(false);
        currentStep + 1 === trace.length
            ? setPlayDisabled(true)
            : setPlayDisabled(false);
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [currentStep]);

    // Use the latest speed to play the animation
    useEffect(() => {
        // if it is playing, replay
        if (isPlaying) {
            pause();
            resume();
        }
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [playSpeed]);

    // It is used to open the speed menu
    const handleClick = (event) => {
        setAnchorEl(event.currentTarget);
    };

    // It is used to adjust speed, triggered when close the speed menu
    const handleClose = (event) => {
        const value = event.nativeEvent.target.value / 4;
        if (!isNaN(value)) {
            setPlaySpeed(value);
        }
        setAnchorEl(null);
    };

    const handleSliderChange = (event, newValue) => {
        if (isPlaying) {
            pause();
        }
        console.log(newValue);

        const item = trace[newValue];
        setCurrentStep(newValue);
        setBars(item);
    };

    // It is used to clean timeouts to pause the animation
    const clearTimeouts = () => {
        timeOutIds.forEach((timeoutId) => clearTimeout(timeoutId));
        setTimeOutIds([]);
    };

    // Main method, used to generate a time line based on the speed
    // change bars after a time interval
    const run = (trace) => {
        // The current bars are the [0] of trace, to make the animation start without delay
        // here we begin the animation from [1] of trace
        // it's also suitable for pause, step forward and backwar
        const subTrace = trace.slice(1);
        const timeoutIds = [];
        // a time interval unit
        const timer = 1000 / playSpeed;

        // Set a timeout for each item in the trace
        subTrace.forEach((item, i) => {
            let timeoutId = setTimeout(
                (item) => {
                    // update the current step
                    setCurrentStep((prevStep) =>
                        i === trace.length - 1 ? prevStep : prevStep + 1
                    );
                    // update bars to be animated
                    setBars(item);
                    i === subTrace.length - 1
                        ? setIsPlaying(false)
                        : setIsPlaying(true);
                },
                i * timer, //time interval
                item
            );
            timeoutIds.push(timeoutId);
        });

        // Clear timeouts upon completion
        let timeoutId = setTimeout(clearTimeouts, trace.length * timer);
        timeoutIds.push(timeoutId);
        setTimeOutIds(timeoutIds);
    };

    // To pause the animation
    const pause = () => {
        setIsPlaying(false);
        clearTimeouts();
    };

    // To resume the animation
    const resume = () => {
        setIsPlaying(true);
        const newtrace = trace.slice(currentStep);
        run(newtrace);
    };

    // Go to next step and pause
    const stepForward = () => {
        if (currentStep < trace.length - 1) {
            pause();
            const item = trace[currentStep + 1];
            setCurrentStep((prevStep) => prevStep + 1);
            setBars(item);
        }
    };

    // Go to the previous step and pause
    const stepBackward = () => {
        if (currentStep > 0) {
            pause();
            const item = trace[currentStep - 1];
            setCurrentStep((prevStep) => prevStep - 1);
            setBars(item);
        }
    };

    const handleResetClick = () => {
        pause();
        setCurrentStep(0);
        setBars(trace[0]);
    };

    const animationControlProps = {
        handleResetClick,
        stepForward,
        stepBackward,
        pause,
        resume,
        isPlaying,
        playDisabled,
        backwardDisabled,
        handleClick,
        handleClose,
        anchorEl,
        playSpeed,
        trace,
    };

    return (
        <div className={classes.root}>
            <div className={classes.title}>{title}</div>
            <Card className={classes.cardOne}>
                <div className={classes.bars}>
                    {bars.map((background) => (
                        <motion.li
                            key={background.key} // each bar's identification
                            layout
                            transition={spring}
                            style={background}
                            className={classes.bar}
                            animate={{
                                backgroundColor: background.backgroundColor,
                                y: background.y,
                            }}
                        >
                            <div className={classes.barNumber}>
                                {background.value}
                            </div>
                            <div
                                style={{
                                    marginTop: background.height - 17,
                                    fontSize: "16px",
                                    textAlign: "center",
                                    fontWeight: "600",
                                }}
                            >
                                {background.isPivot ? <ArrowDropUpIcon style={{ marginBottom: "-18px" }} /> : ""}
                            </div>
                        </motion.li>
                    ))}
                </div>
                <p className={classes.explanation}>
                    {description[currentStep]}
                </p>

                <div className={classes.slider}>
                    <AnimationSlider
                        width={width}
                        step={1}
                        max={trace.length - 1}
                        handleChange={handleSliderChange}
                        value={currentStep}
                        display="none"
                    />
                </div>
                <div className={classes.control}>
                    <AnimationControl {...animationControlProps} />
                </div>
            </Card>

            <Card className={classes.cardTwo}>
                <CardContent>
                    <div style={{ lineHeight: "25px" }}>
                        <Code blockNum={blockNums[currentStep]} />
                    </div>
                </CardContent>
            </Card>
        </div>
    );
};
