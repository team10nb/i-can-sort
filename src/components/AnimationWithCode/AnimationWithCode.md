### **Developer**

Ruizi Han

Reference: MATERIAL-UI (https://material-ui.com)

### **Introduction**

Animation component which contains a control bar, a slider, a description box and pseudocode which can be highlighted.

### **Import**

```html
import "@fontsource/roboto";
import * as React from "react";
import { motion } from "framer-motion";
import { useState, useEffect } from "react";
import AnimationSlider from "../AnimationSlider/AnimationSlider";
import AnimationControl from "../AnimationControl/AnimationControl";
import ArrowDropUpIcon from '@material-ui/icons/ArrowDropUp';
import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
```

### **Props**

| Name             | Type   | Default | Description                                                   |
| ---------------- | ------ | ------- | ------------------------------------------------------------- |
| trace         | object | null    | Indicate the sorting process                             |
| description         | [string] | []    | Contain array of descriptions                             |
| width | number | 0       | The width of the slider          |
| blockNums | [number] | [] | The array of highlighted pseudocode block number |
| useStyles | style | {} | The css of this component |
| Code | Code |  | Pseudocode blocks |
| title | string | "" | The title |

### **Demo**

```jsx
import { useStyles } from "../../scenes/subpages/Tutorial/Loop/LoopExample/Style";
import { makeStyles } from "@material-ui/core/styles";


const Code = (props) => {
    const { blockNum } = props;

    const useStyles = makeStyles((theme) => ({
        background: {
            backgroundColor: "#FEE5D4",
            borderRadius: 5,
            marginTop: -6,
            letterSpacing: 0.5,
            fontFamily: "Space Mono",
            lineHeight: "22px",
            fontSize: "12px",
            paddingLeft: "5px",
        },
        noBackground: {
            marginTop: -6,
            letterSpacing: 0.5,
            fontFamily: "Space Mono",
            lineHeight: "22px",
            fontSize: "12px",
            paddingLeft: "5px",
        },
    }));
    
    const classes = useStyles();

    const terminableCode =
        <div>
            <pre className={blockNum === 1 ? classes.background : classes.noBackground}>
                {`Algorithm: SwapFourTimes(Arr)`}
            </pre>
            <pre className={classes.noBackground}>
                {`  i = 0`}
            </pre>
            <pre className={blockNum === 2 ? classes.background : classes.noBackground}>
                {`  for i from 0 to 3 do`}
            </pre>
            <pre className={blockNum === 3 ? classes.background : classes.noBackground}>
                {`    swap(Arr[0], Arr[1])`}
            </pre>
            <pre className={blockNum === 4 ? classes.background : classes.noBackground}>
                {`  end for`}
            </pre>
        </div>

    return (
        <div>{terminableCode}</div>
    )
}

const trace = [];
const traceUnit1 =
        [
            { value: 5, height: 50, backgroundColor: "#00BFFF", key: 0, y: 0 },
            { value: 8, height: 80, backgroundColor: "#00BFFF", key: 1, y: 0 },
        ];

    const traceUnit2 =
        [
            { value: 5, height: 50, backgroundColor: "#FFD700", key: 0, y: 0 },
            { value: 8, height: 80, backgroundColor: "#FFD700", key: 1, y: 0 },
        ];
    
    const traceUnit3 =
        [
            { value: 8, height: 80, backgroundColor: "#00BFFF", key: 1, y: 0 },
            { value: 5, height: 50, backgroundColor: "#00BFFF", key: 0, y: 0 },
        ];

    const traceUnit4 =
        [
            { value: 8, height: 80, backgroundColor: "#FFD700", key: 1, y: 0 },
            { value: 5, height: 50, backgroundColor: "#FFD700", key: 0, y: 0 },
        ];
    trace.push(traceUnit1, traceUnit2, traceUnit3, traceUnit4, traceUnit1, traceUnit2, traceUnit3, traceUnit4, traceUnit1, traceUnit1);

    const blockNums = [1, 2, 3, 2, 3, 2, 3, 2, 3, 4];

    const description = [
        "i = 0",
        "i = 0",
        "i = 0",
        "i = 1",
        "i = 1",
        "i = 2",
        "i = 2",
        "i = 3",
        "i = 3",
        "i = 4, jump out of the loop",
    ];

<AnimationWithCode
    trace={trace}
    description={description}
    width={320}
    blockNums={blockNums}
    useStyles={useStyles}
    Code={Code}
    title="Demo"
/>
```