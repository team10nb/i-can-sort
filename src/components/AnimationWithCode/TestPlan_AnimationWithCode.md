## Test Plan
--------
<br>


#### Component: AnimationWithCode

|Test Name|Description|Operation|Expected Outcome|
|----|------|----------------|------------|
|should call AnimationSlider||||
|should call AnimationControl||||
| should match Snapshot | Compare new snapshot with old one. |  | Snapshots are the same. |
