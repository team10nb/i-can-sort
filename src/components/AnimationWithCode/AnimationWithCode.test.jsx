/*
    Author: Ruizi Han
*/

import { render } from "@testing-library/react";
import { makeStyles } from "@material-ui/core/styles";
import AnimationWithCode from "../AnimationWithCode/AnimationWithCode";
import * as AnimationSliders from "../AnimationSlider/AnimationSlider";
import * as AnimationControls from "../AnimationControl/AnimationControl";

const trace = [
    [
        { height: "50", backgroundColor: "#FF008C", key: '1' },
        { height: "30", backgroundColor: "#D309E1", key: '2' },
        { height: "90", backgroundColor: "#9C1AFF", key: '3' },
        { height: "10", backgroundColor: "#7700FF", key: '4' },
        { height: "20", backgroundColor: "#66AA00", key: '5' },
    ],[
        { height: "50", backgroundColor: "#FF008C", key: '1' },
        { height: "30", backgroundColor: "#D309E1", key: '2' },
        { height: "90", backgroundColor: "#9C1AFF", key: '3' },
        { height: "10", backgroundColor: "#7700FF", key: '4' },
        { height: "20", backgroundColor: "#66AA00", key: '5' },
    ],

];

const description = ["", ""];

const blockNums = [1, 2];

const useStyles = makeStyles((theme) => ({
    root: {
        '& > * + *': {
            marginTop: theme.spacing(3),
        },
        width: 380,
    },
    bars: {
        listStyle: "none",
        padding: 0,
        margin: 0,
        position: "relative",
        display: "flex",
        flexWrap: "wrap-reverse",
        justifyContent: "center",
    },
    bar: {
        listStyle: "none",
        padding: 0,
        margin: 0,
        boxShadow: "0px 0px 2px 2px #88888833",
        borderRadius: "10px",
        marginTop: 50,
        marginBottom: "-5px",
        marginRight: "15px",
        marginLeft: "15px",
        width: "30px",
        height: "140px",
        fontSize: "0.1em",
        justifyContent: "center",
    },
    slider: {
        marginLeft: 28,
        marginTop: 10,
    },
}));

const Code = (props) => {
    const { blockNum } = props;

    const useStyles = makeStyles((theme) => ({
        background: {
            backgroundColor: "#FEE5D4",
            borderRadius: 5,
            marginTop: -6,
            letterSpacing: 0.5,
            fontFamily: "Space Mono",
            lineHeight: "22px",
            fontSize: "12px",
            paddingLeft: "5px",
        },
        noBackground: {
            marginTop: -6,
            letterSpacing: 0.5,
            fontFamily: "Space Mono",
            lineHeight: "22px",
            fontSize: "12px",
            paddingLeft: "5px",
        },
    }));
    
    const classes = useStyles();

    const testCode =
        <div>
            <pre className={blockNum === 1 ? classes.background : classes.noBackground}>
                {`test`}
            </pre>
            <pre className={blockNum === 2 ? classes.background : classes.noBackground}>
                {`test`}
            </pre>
        </div>

    return (
        <div>{testCode}</div>
    )
}

const props = {
    trace,
    description,
    width: 320,
    blockNums,
    useStyles,
    Code,
    title: "For Loop",
};

test('should call AnimationSlider', () => {
    jest.restoreAllMocks();
    const sliderSpy = jest.spyOn(AnimationSliders, "default").mockImplementation(() => <div>mockAnimationSlider</div>);
    const content = render(<AnimationWithCode {...props}/>);
    content.getByText("mockAnimationSlider");
    expect(sliderSpy).toBeCalledTimes(1);
});

test('should call AnimationControl', () => {
    jest.restoreAllMocks();
    const controlSpy = jest.spyOn(AnimationControls, "default").mockImplementation(() => <div>mockAnimationControl</div>);

    const {getByText} = render(<AnimationWithCode {...props}/>);
    getByText("mockAnimationControl");
    expect(controlSpy).toBeCalledTimes(1);
});

test('should match snapshot', () => {
    jest.restoreAllMocks();
    // jest.resetModules();
    const content = render(<AnimationWithCode {...props}/>);

    expect(content).toMatchSnapshot();
});