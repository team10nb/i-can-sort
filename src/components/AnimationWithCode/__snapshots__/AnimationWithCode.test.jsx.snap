// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="makeStyles-root-32"
      >
        <div>
          For Loop
        </div>
        <div
          class="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
        >
          <div
            class="makeStyles-bars-33"
          >
            <li
              class="makeStyles-bar-34"
              style="background-color: rgb(255, 0, 140); transform: none;"
            >
              <div />
              <div
                style="margin-top: 33px; font-size: 16px; text-align: center; font-weight: 600;"
              />
            </li>
            <li
              class="makeStyles-bar-34"
              style="background-color: rgb(211, 9, 225); transform: none;"
            >
              <div />
              <div
                style="margin-top: 13px; font-size: 16px; text-align: center; font-weight: 600;"
              />
            </li>
            <li
              class="makeStyles-bar-34"
              style="background-color: rgb(156, 26, 255); transform: none;"
            >
              <div />
              <div
                style="margin-top: 73px; font-size: 16px; text-align: center; font-weight: 600;"
              />
            </li>
            <li
              class="makeStyles-bar-34"
              style="background-color: rgb(119, 0, 255); transform: none;"
            >
              <div />
              <div
                style="margin-top: -7px; font-size: 16px; text-align: center; font-weight: 600;"
              />
            </li>
            <li
              class="makeStyles-bar-34"
              style="background-color: rgb(102, 170, 0); transform: none;"
            >
              <div />
              <div
                style="margin-top: 3px; font-size: 16px; text-align: center; font-weight: 600;"
              />
            </li>
          </div>
          <p />
          <div
            class="makeStyles-slider-35"
          >
            <div
              class="makeStyles-root-36"
            >
              <span
                class="MuiSlider-root WithStyles(ForwardRef(Slider))-root-37 MuiSlider-colorPrimary"
              >
                <span
                  class="MuiSlider-rail WithStyles(ForwardRef(Slider))-rail-42"
                />
                <span
                  class="MuiSlider-track WithStyles(ForwardRef(Slider))-track-41"
                  style="left: 0%; width: 0%;"
                />
                <input
                  type="hidden"
                  value="0"
                />
                <span
                  class="MuiSlider-mark WithStyles(ForwardRef(Slider))-mark-43 MuiSlider-markActive"
                  data-index="0"
                  style="left: 0%;"
                />
                <span
                  class="MuiSlider-mark WithStyles(ForwardRef(Slider))-mark-43"
                  data-index="1"
                  style="left: 100%;"
                />
                <span
                  aria-labelledby="Progress"
                  aria-orientation="horizontal"
                  aria-valuemax="1"
                  aria-valuemin="0"
                  aria-valuenow="0"
                  class="MuiSlider-thumb WithStyles(ForwardRef(Slider))-thumb-38 MuiSlider-thumbColorPrimary PrivateValueLabel-thumb-44"
                  data-index="0"
                  role="slider"
                  style="left: 0%;"
                  tabindex="0"
                >
                  <span
                    class="PrivateValueLabel-offset-46 MuiSlider-valueLabel WithStyles(ForwardRef(Slider))-valueLabel-40"
                  >
                    <span
                      class="PrivateValueLabel-circle-47"
                    >
                      <span
                        class="PrivateValueLabel-label-48"
                      >
                        0
                      </span>
                    </span>
                  </span>
                </span>
              </span>
            </div>
          </div>
          <div>
            <div
              style="position: relative; left: 50%; transform: translate(-50%); width: 240px;"
            >
              <span
                class="makeStyles-customTooltip-49"
                title="Reset"
              >
                <button
                  class="MuiButtonBase-root MuiIconButton-root"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                      focusable="false"
                      style="color: grey;"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M12 5V1L7 6l5 5V7c3.31 0 6 2.69 6 6s-2.69 6-6 6-6-2.69-6-6H4c0 4.42 3.58 8 8 8s8-3.58 8-8-3.58-8-8-8z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </span>
              <button
                class="MuiButtonBase-root MuiIconButton-root Mui-disabled Mui-disabled"
                disabled=""
                tabindex="-1"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                    focusable="false"
                    style="color: grey;"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M6 6h2v12H6zm3.5 6l8.5 6V6z"
                    />
                  </svg>
                </span>
              </button>
              <button
                class="MuiButtonBase-root MuiIconButton-root"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 14.5v-9l6 4.5-6 4.5z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                class="MuiButtonBase-root MuiIconButton-root"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                    focusable="false"
                    style="color: grey;"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M6 18l8.5-6L6 6v12zM16 6v12h2V6h-2z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <div
                class="makeStyles-root-50"
              >
                <button
                  aria-controls="simple-menu"
                  aria-haspopup="true"
                  class="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-button-52"
                  style="min-width: 40px;"
                  tabindex="0"
                  title="Speed"
                  type="button"
                >
                  <span
                    class="MuiButton-label"
                  >
                    1x
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
        >
          <div
            class="MuiCardContent-root"
          >
            <div
              style="line-height: 25px;"
            >
              <div>
                <div>
                  <pre
                    class="makeStyles-background-54"
                  >
                    test
                  </pre>
                  <pre
                    class="makeStyles-noBackground-55"
                  >
                    test
                  </pre>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      aria-hidden="true"
      class="MuiPopover-root makeStyles-menu-51"
      id="simple-menu"
      role="presentation"
      style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px; visibility: hidden;"
    >
      <div
        data-test="sentinelStart"
        tabindex="0"
      />
      <div
        class="MuiPaper-root MuiMenu-paper MuiPopover-paper MuiPaper-elevation8 MuiPaper-rounded"
        style="opacity: 0; transform: scale(0.75, 0.5625); visibility: hidden;"
        tabindex="-1"
      >
        <ul
          class="MuiList-root MuiMenu-list MuiList-padding"
          role="menu"
          tabindex="-1"
        >
          <li
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root WithStyles(ForwardRef(MenuItem))-root-56 MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button"
            role="menuitem"
            tabindex="0"
            value="1"
          >
            0.25x
            <span
              class="MuiTouchRipple-root"
            />
          </li>
          <li
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root WithStyles(ForwardRef(MenuItem))-root-56 MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button"
            role="menuitem"
            tabindex="-1"
            value="2"
          >
            0.5x
            <span
              class="MuiTouchRipple-root"
            />
          </li>
          <li
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root WithStyles(ForwardRef(MenuItem))-root-56 MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button"
            role="menuitem"
            tabindex="-1"
            value="4"
          >
            1x
            <span
              class="MuiTouchRipple-root"
            />
          </li>
          <li
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root WithStyles(ForwardRef(MenuItem))-root-56 MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button"
            role="menuitem"
            tabindex="-1"
            value="8"
          >
            2x
            <span
              class="MuiTouchRipple-root"
            />
          </li>
          <li
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root WithStyles(ForwardRef(MenuItem))-root-56 MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button"
            role="menuitem"
            tabindex="-1"
            value="16"
          >
            4x
            <span
              class="MuiTouchRipple-root"
            />
          </li>
        </ul>
      </div>
      <div
        data-test="sentinelEnd"
        tabindex="0"
      />
    </div>
  </body>,
  "container": <div>
    <div
      class="makeStyles-root-32"
    >
      <div>
        For Loop
      </div>
      <div
        class="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div
          class="makeStyles-bars-33"
        >
          <li
            class="makeStyles-bar-34"
            style="background-color: rgb(255, 0, 140); transform: none;"
          >
            <div />
            <div
              style="margin-top: 33px; font-size: 16px; text-align: center; font-weight: 600;"
            />
          </li>
          <li
            class="makeStyles-bar-34"
            style="background-color: rgb(211, 9, 225); transform: none;"
          >
            <div />
            <div
              style="margin-top: 13px; font-size: 16px; text-align: center; font-weight: 600;"
            />
          </li>
          <li
            class="makeStyles-bar-34"
            style="background-color: rgb(156, 26, 255); transform: none;"
          >
            <div />
            <div
              style="margin-top: 73px; font-size: 16px; text-align: center; font-weight: 600;"
            />
          </li>
          <li
            class="makeStyles-bar-34"
            style="background-color: rgb(119, 0, 255); transform: none;"
          >
            <div />
            <div
              style="margin-top: -7px; font-size: 16px; text-align: center; font-weight: 600;"
            />
          </li>
          <li
            class="makeStyles-bar-34"
            style="background-color: rgb(102, 170, 0); transform: none;"
          >
            <div />
            <div
              style="margin-top: 3px; font-size: 16px; text-align: center; font-weight: 600;"
            />
          </li>
        </div>
        <p />
        <div
          class="makeStyles-slider-35"
        >
          <div
            class="makeStyles-root-36"
          >
            <span
              class="MuiSlider-root WithStyles(ForwardRef(Slider))-root-37 MuiSlider-colorPrimary"
            >
              <span
                class="MuiSlider-rail WithStyles(ForwardRef(Slider))-rail-42"
              />
              <span
                class="MuiSlider-track WithStyles(ForwardRef(Slider))-track-41"
                style="left: 0%; width: 0%;"
              />
              <input
                type="hidden"
                value="0"
              />
              <span
                class="MuiSlider-mark WithStyles(ForwardRef(Slider))-mark-43 MuiSlider-markActive"
                data-index="0"
                style="left: 0%;"
              />
              <span
                class="MuiSlider-mark WithStyles(ForwardRef(Slider))-mark-43"
                data-index="1"
                style="left: 100%;"
              />
              <span
                aria-labelledby="Progress"
                aria-orientation="horizontal"
                aria-valuemax="1"
                aria-valuemin="0"
                aria-valuenow="0"
                class="MuiSlider-thumb WithStyles(ForwardRef(Slider))-thumb-38 MuiSlider-thumbColorPrimary PrivateValueLabel-thumb-44"
                data-index="0"
                role="slider"
                style="left: 0%;"
                tabindex="0"
              >
                <span
                  class="PrivateValueLabel-offset-46 MuiSlider-valueLabel WithStyles(ForwardRef(Slider))-valueLabel-40"
                >
                  <span
                    class="PrivateValueLabel-circle-47"
                  >
                    <span
                      class="PrivateValueLabel-label-48"
                    >
                      0
                    </span>
                  </span>
                </span>
              </span>
            </span>
          </div>
        </div>
        <div>
          <div
            style="position: relative; left: 50%; transform: translate(-50%); width: 240px;"
          >
            <span
              class="makeStyles-customTooltip-49"
              title="Reset"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                    focusable="false"
                    style="color: grey;"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 5V1L7 6l5 5V7c3.31 0 6 2.69 6 6s-2.69 6-6 6-6-2.69-6-6H4c0 4.42 3.58 8 8 8s8-3.58 8-8-3.58-8-8-8z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </span>
            <button
              class="MuiButtonBase-root MuiIconButton-root Mui-disabled Mui-disabled"
              disabled=""
              tabindex="-1"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                  focusable="false"
                  style="color: grey;"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M6 6h2v12H6zm3.5 6l8.5 6V6z"
                  />
                </svg>
              </span>
            </button>
            <button
              class="MuiButtonBase-root MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 14.5v-9l6 4.5-6 4.5z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButtonBase-root MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                  focusable="false"
                  style="color: grey;"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M6 18l8.5-6L6 6v12zM16 6v12h2V6h-2z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
            <div
              class="makeStyles-root-50"
            >
              <button
                aria-controls="simple-menu"
                aria-haspopup="true"
                class="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-button-52"
                style="min-width: 40px;"
                tabindex="0"
                title="Speed"
                type="button"
              >
                <span
                  class="MuiButton-label"
                >
                  1x
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div
          class="MuiCardContent-root"
        >
          <div
            style="line-height: 25px;"
          >
            <div>
              <div>
                <pre
                  class="makeStyles-background-54"
                >
                  test
                </pre>
                <pre
                  class="makeStyles-noBackground-55"
                >
                  test
                </pre>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
